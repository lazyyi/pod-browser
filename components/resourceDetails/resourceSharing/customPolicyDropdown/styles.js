/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { createStyles } from "@solid/lit-prism-patterns";

const SETTINGS_ICON_COLOR = "#404040";
const SETTINGS_BACKGROUND_COLOR = "#F5F5F5";

export default function styles(theme) {
  return createStyles(theme, ["icons"], {
    dropdown: {
      display: "flex",
      border: `1px solid ${theme.palette.grey.A100}`,
      borderRadius: theme.shape.borderRadius,
      marginBottom: theme.spacing(1.2),
      padding: theme.spacing(1.2),
      fontWeight: theme.typography.body.fontWeight,
      fontFamily: theme.typography.body.fontFamily,
      backgroundColor: theme.palette.background.paper,
    },
    menuItem: {
      fontWeight: theme.typography.body.fontWeight,
      fontFamily: theme.typography.body.fontFamily,
    },
    manuItemText: {
      padding: 0,
    },
    title: {
      fontSize: theme.typography.h4.fontSize,
      fontWeight: theme.typography.fontWeightBold,
    },
    headerContainer: {
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      marginBottom: theme.spacing(1.6),
    },
    textContainer: {
      width: "100%",
      flexDirection: "column",
    },
    description: {
      "&& p": {
        padding: 0,
        margin: 0,
        whiteSpace: "break-spaces",
      },
      fontSize: "0.8125rem",
    },
    titleAndButtonContainer: {
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between",
    },
    "icon-custom-policies": {
      width: "1.5rem",
      height: "1.5rem",
      fontStretch: "unset",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      color: SETTINGS_ICON_COLOR,
      backgroundColor: SETTINGS_BACKGROUND_COLOR,
      padding: theme.spacing(1),
      borderRadius: "50%",
      marginRight: theme.spacing(1.6),
      fontSize: theme.typography.h1.fontSize,
    },
    iconCaret: {
      fontSize: "34px",
      top: "25%",
      color: theme.palette.text.primary,
      marginRight: "1rem",
    },
  });
}
