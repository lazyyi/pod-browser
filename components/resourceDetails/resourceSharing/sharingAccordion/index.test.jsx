/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import { waitFor, within } from "@testing-library/react";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import mockPersonContact from "../../../../__testUtils/mockPersonContact";
import mockAddressBook from "../../../../__testUtils/mockAddressBook";
import SharingAccordion from "./index";
import { TESTCAFE_ID_AGENT_ACCESS_TABLE } from "../agentAccessTable";
import { TESTCAFE_ID_ADVANCED_SHARING_BUTTON } from "../advancedSharingButton";
import useContacts from "../../../../src/hooks/useContacts";
import useAddressBook from "../../../../src/hooks/useAddressBook";
import { TESTCAFE_ID_CUSTOM_POLICY_DROPDOWN } from "../customPolicyDropdown";

jest.mock("../../../../src/hooks/useContacts");
const mockedUseContacts = useContacts;

jest.mock("../../../../src/hooks/useAddressBook");
const mockedUseAddressBook = useAddressBook;

describe("SharingAccordion", () => {
  beforeEach(() => {
    mockedUseAddressBook.mockReturnValue({ data: mockAddressBook() });
    mockedUseContacts.mockReturnValue({
      data: [
        mockPersonContact(
          mockAddressBook(),
          "https://example.org/contacts/Person/1234/",
          "Example 1"
        ),
        mockPersonContact(
          mockAddressBook(),
          "https://example.org/contacts/Person/3456/",
          "Example 2"
        ),
      ],
    });
  });
  test("it renders three lists of empty permissions for editors, viewers and blocked, and an Advanced Sharing button", () => {
    const { asFragment, queryAllByTestId, queryByTestId } = renderWithTheme(
      <SharingAccordion />
    );
    expect(queryAllByTestId(TESTCAFE_ID_AGENT_ACCESS_TABLE)).toHaveLength(3);
    expect(queryByTestId(TESTCAFE_ID_ADVANCED_SHARING_BUTTON)).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
  test("after editing view & add policy, the permissions for the edited policy are displayed", () => {
    const {
      getByTestId,
      queryAllByTestId,
      queryByText,
      queryAllByRole,
    } = renderWithTheme(<SharingAccordion />);
    const advancedSharingButton = getByTestId(
      TESTCAFE_ID_ADVANCED_SHARING_BUTTON
    );
    userEvent.click(advancedSharingButton);
    const checkboxes = queryAllByRole("checkbox");
    userEvent.click(checkboxes[0]);
    userEvent.click(checkboxes[1]);
    const submitButton = getByTestId("submit-webids-button");
    userEvent.click(submitButton);
    waitFor(() => {
      const renderedTables = queryAllByTestId(TESTCAFE_ID_AGENT_ACCESS_TABLE);
      expect(renderedTables).toHaveLength(4);
      const title = queryByText("View & Add");
      const description = queryByText(
        "Can view and add new content but cannot edit or delete existing content"
      );
      expect(title).not.toBeNull();
      expect(description).not.toBeNull();
    });
  });
  test("after changing the selected custom policy in dropdown and editing it, the permissions for the edited policy are displayed", () => {
    const {
      getByTestId,
      queryAllByTestId,
      queryByText,
      queryAllByRole,
    } = renderWithTheme(<SharingAccordion />);
    const advancedSharingButton = getByTestId(
      TESTCAFE_ID_ADVANCED_SHARING_BUTTON
    );
    userEvent.click(advancedSharingButton);
    const dropdown = getByTestId(TESTCAFE_ID_CUSTOM_POLICY_DROPDOWN);
    userEvent.click(within(dropdown).getByRole("button"));
    expect(queryAllByRole("option")).toHaveLength(3);
    const editOnlyOption = queryAllByRole("option")[1];
    userEvent.click(editOnlyOption);
    const checkboxes = queryAllByRole("checkbox");
    userEvent.click(checkboxes[0]);
    userEvent.click(checkboxes[1]);
    const submitButton = getByTestId("submit-webids-button");
    userEvent.click(submitButton);
    waitFor(() => {
      const renderedTables = queryAllByTestId(TESTCAFE_ID_AGENT_ACCESS_TABLE);
      expect(renderedTables).toHaveLength(4);
      const title = queryByText("Edit Only");
      const description = queryByText(
        "Can edit existing content but cannot view or delete existing content"
      );
      expect(title).not.toBeNull();
      expect(description).not.toBeNull();
    });
  });
});
