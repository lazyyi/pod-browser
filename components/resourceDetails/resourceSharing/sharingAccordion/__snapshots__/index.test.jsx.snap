// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharingAccordion it renders three lists of empty permissions for editors, viewers and blocked, and an Advanced Sharing button 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
    data-testid="agent-access-table"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-editor PodBrowser-icon PodBrowser-iconEditor"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Editors
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--text"
            data-testid="add-agent-button"
            iconbefore="add"
          >
            <span
              class="PodBrowser-icon-add PodBrowser-button__icon PodBrowser-button__icon--before"
            />
            <span
              class="PodBrowser-button__text"
            >
              Edit Editors
            </span>
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view, edit and delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No editors
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
    data-testid="agent-access-table"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-view PodBrowser-icon PodBrowser-iconViewer"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Viewers
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--text"
            data-testid="add-agent-button"
            iconbefore="add"
          >
            <span
              class="PodBrowser-icon-add PodBrowser-button__icon PodBrowser-button__icon--before"
            />
            <span
              class="PodBrowser-button__text"
            >
              Edit Viewers
            </span>
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Can 
            </b>
            view but
            <b>
               cannot 
            </b>
            edit or delete this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No viewers
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="PodBrowser-root PodBrowser-root PodBrowser-accordion PodBrowser-expanded PodBrowser-rounded PodBrowser-rounded PodBrowser-elevation1 PodBrowser-rounded"
    data-testid="agent-access-table"
  >
    <div
      class="PodBrowser-headerContainer"
    >
      <i
        class="PodBrowser-icon-block PodBrowser-icon PodBrowser-iconBlocked"
      />
      <div
        class="PodBrowser-textContainer"
      >
        <div
          class="PodBrowser-titleAndButtonContainer"
        >
          <p
            class="PodBrowser-title"
          >
            Blocked
          </p>
          <button
            class="PodBrowser-button PodBrowser-button--text"
            data-testid="add-agent-button"
            iconbefore="add"
          >
            <span
              class="PodBrowser-icon-add PodBrowser-button__icon PodBrowser-button__icon--before"
            />
            <span
              class="PodBrowser-button__text"
            >
              Edit Blocked
            </span>
          </button>
        </div>
        <span
          class="PodBrowser-description"
        >
          <p>
            <b>
              Cannot 
            </b>
            view this resource
          </p>
        </span>
      </div>
    </div>
    <div
      class="PodBrowser-container PodBrowser-entered"
      style="min-height: 0px;"
    >
      <div
        class="PodBrowser-wrapper"
      >
        <div
          class="PodBrowser-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="PodBrowser-permissionsContainer"
            >
              <span
                class="PodBrowser-emptyStateTextContainer"
              >
                <p>
                  No one is blocked
                </p>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="PodBrowser-sharingButtonContainer"
  >
    <button
      class="PodBrowser-button PodBrowser-button--text PodBrowser-sharingButton"
      data-testid="advanced-sharing-button"
    >
      <span
        class="PodBrowser-button__text"
      >
        <i
          class="PodBrowser-icon-settings PodBrowser-icon"
        />
        Advanced Sharing
      </span>
    </button>
  </div>
</DocumentFragment>
`;
